rfc/rfc-81/rfc-81.md

```diff
@@ -0,0 +1,108 @@
+<!--
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+# RFC-81: Log Compaction with Merge Sort
+
+## Proposers
+- @usberkeley
+
+## Approvers
+- @danny0405
+
+## Status
+JIRA: https://issues.apache.org/jira/browse/HUDI-8033
+
+## Abstract
+Add lightweight LogCompaction to improve the writing performance of the write side, and improve the query performance of the read side (Spark/Presto, etc.) in some scenarios without having to wait for heavy and time-consuming operations such as Compaction or Clustering.
+
+## Background
+The previous LogCompaction mainly merged log files through HoodieMergedLogRecordScanner, and used ExternalSpillableMap internally to achieve record merging, which resulted in performance loss of writing to disk.
+LogCompaction with Merge Sort is introduced to achieve lightweight minor compaction by merging records through N-way streaming of ordered data, thus improving the writing performance of the write side. At the same time, thanks to the ordered data, the query performance on the read side can be improved when the primary key is met.
```

Comment:
Hi, How does the solution improve both write and read performance?

Reply:
The write performance should have some regression because of the sort procedure, for query performance boost, you mean we sort the inputs by primary key? Can the sort key be customized?

Reply:
> The write performance should have some regression because of the sort procedure, for query performance boost, you mean we sort the inputs by primary key? Can the sort key be customized?

Customized sorting seems hard to be implemented during this merge-sort process, because in this process, we need to do record-combine based on the arrangement of primary keys together. cc @usberkeley Will the problems I mentioned occur in your implementation?

Reply:
> > The write performance should have some regression because of the sort procedure, for query performance boost, you mean we sort the inputs by primary key? Can the sort key be customized?
> 
> Customized sorting seems hard to be implemented during this merge-sort process, because in this process, we need to do record-combine based on the arrangement of primary keys together. cc @usberkeley Will the problems I mentioned occur in your implementation?

As you mentioned, custom sorting is not supported

Reply:
> The write performance should have some regression because of the sort procedure, for query performance boost, you mean we sort the inputs by primary key? Can the sort key be customized?

1）The sort key cannot be customized
2）The write performance will be slightly degraded in DeltaCommit, but Merge Sort avoids the overhead of ExternalSpillableMap writing to disk. Our internal preliminary tests show that data freshness and write throughput are greatly improved.

Reply:
Okay. I benchmarked N way sort merge once vs current hash merge (what we call ExternalSpillableMap). sort merge can be expensive when N is much higher.. Might be worth adding some microbenchmarks to understand. I am overall in favor of adding a sort based merge though. 

https://github.com/vinothchandar/hudi/commit/9050dd1317851bc395aef51ab60d761f74c0529d#diff-a7143db9299e96f8b0dc010a2dcb530318ddc3b30ce38f79881afebea602ad2c